### PR TITLE
Python: Fix agent option merge to support dict-defined tools

### LINKED ISSUE #4303: Python: Agent option merge drops dict-defined tools
### Description
`_merge_options` drops dict-defined tools when merging default and runtime options.

The merge logic de-duplicates tools by `getattr(tool, "name", None)`. For dict tools, that value is `None`, so runtime dict tools are incorrectly treated as duplicates and discarded.

### Code Sample
```python
from agent_framework._agents import _merge_options

base = {
    "tools": [
        {"type": "function", "function": {"name": "tool_a"}},
    ]
}
override = {
    "tools": [
        {"type": "function", "function": {"name": "tool_b"}},
    ]
}

merged = _merge_options(base, override)
print(merged["tools"])
# Actual: only tool_a
# Expected: tool_a and tool_b
```

### Error Messages / Stack Traces
No exception is raised; the runtime tool is silently dropped.

### Package Versions
agent-framework-core: source checkout at commit `02ba27493`

### Python Version
Python 3.13.5

### Additional Context
This impacts hosted/dict-style tool definitions passed via options.

### PR DESCRIPTION
**Pull request by moonbox3**
### Motivation and Context

When tools were defined as plain dicts (the `{"type": "function", "function": {"name": ...}}` format accepted by OpenAI-compatible APIs), `_merge_options` used `getattr(t, "name", None)` to extract tool names for deduplication, which always returned `None` for dicts — causing all override tools to be silently dropped as false-positive duplicates.

Fixes #4303

### Description

The root cause was that `_merge_options` assumed tools were always objects with a `.name` attribute. For dict-defined tools, the name lives at `tool["function"]["name"]`, so `getattr` returned `None` for every tool, and the `None`-populated `existing_names` set incorrectly matched all incoming tools. The fix introduces a `_get_tool_name` helper that handles both object and dict tool formats, and discards `None` from the existing-names set so that unnamed tools never cause spurious deduplication.

### Contribution Checklist

- [X] The code builds clean without any errors or warnings
- [X] The PR follows the [Contribution Guidelines](https://github.com/microsoft/agent-framework/blob/main/CONTRIBUTING.md)
- [X] All unit tests pass, and I have added new tests where possible
- [ ] **Is this a breaking change?** If yes, add "[BREAKING]" prefix to the title of the PR.

---

> Note: PR autogenerated by moonbox3's agent

### CODE REVIEW

File: python/packages/core/agent_framework/_agents.py [hunk 1]
```diff
@@ -81,6 +81,14 @@
 )
 
 
+def _get_tool_name(tool: Any) -> str | None:
+    """Extract the name from a tool, supporting both objects and dicts."""
+    if isinstance(tool, dict):
+        name: str | None = tool.get("function", {}).get("name")
```
Comment (moonbox3) [inline python/packages/core/agent_framework/_agents.py:87]: If `tool["function"]` exists but is `None` (or another non-dict), `tool.get("function", {})` returns `None` (the default is only used when the key is *missing*), and the chained `.get("name")` raises `AttributeError`. Consider a guard for this edge case.

```suggestion
    if isinstance(tool, dict):
        func = tool.get("function")
        if isinstance(func, dict):
            return func.get("name")
        return None
```

File: python/packages/core/tests/core/test_agents.py [hunk 2]
```diff
@@ -880,6 +880,71 @@ def __init__(self, name):
...
+
+
+def test_get_tool_name_no_name():
+    """Test _get_tool_name returns None when no name is available."""
+    assert _get_tool_name({}) is None
+    assert _get_tool_name(42) is None
+
+
 def test_merge_options_logit_bias_merged():
     """Test _merge_options merges logit_bias dicts."""
```
Comment (moonbox3) [inline python/packages/core/tests/core/test_agents.py:949]: Missing test for the `discard(None)` behavior on line 108 of `_agents.py`. When both base and override contain tools without names (e.g., `{"type": "function", "function": {}}`), the discard ensures both are kept rather than the override tool being silently dropped. Add a test after this block to cover that path.

```suggestion
def test_get_tool_name_no_name():
    """Test _get_tool_name returns None when no name is available."""
    assert _get_tool_name({}) is None
    assert _get_tool_name(42) is None


def test_merge_options_nameless_tools_not_dropped():
    """Test _merge_options keeps nameless tools from both base and override."""
    base = {
        "tools": [
            {"type": "function", "function": {}},
        ]
    }
    override = {
        "tools": [
            {"type": "function", "function": {}},
        ]
    }

    result = _merge_options(base, override)

    assert len(result["tools"]) == 2


def test_merge_options_mixed_tool_types():
    """Test _merge_options deduplicates across object and dict tool formats."""

    class MockTool:
        def __init__(self, name: str):
            self.name = name

    base = {"tools": [MockTool("shared_tool")]}
    override = {
        "tools": [
            {"type": "function", "function": {"name": "shared_tool"}},
            {"type": "function", "function": {"name": "new_tool"}},
        ]
    }

    result = _merge_options(base, override)

    names = [_get_tool_name(t) for t in result["tools"]]
    assert names.count("shared_tool") == 1
    assert "new_tool" in names

```

**Comment by markwallace-microsoft** [comment]
<a href="https://github.com/microsoft/agent-framework/blob/11365dc3cb633256feae29772fd9a135b6211e9e/README.md">Python Test Coverage</a><details><summary>Python Test Coverage Report • </summary><table><tr><th>File</th><th>Stmts</th><th>Miss</th><th>Cover</th><th>Missing</th></tr><tbody><tr><td colspan="5"><b>packages/core/agent_framework</b></td></tr><tr><td>&nbsp; &nbsp;<a href="https://github.com/microsoft/agent-framework/blob/11365dc3cb633256feae29772fd9a135b6211e9e/packages/core/agent_framework/_agents.py">_agents.py</a></td><td>337</td><td>43</td><td>87%</td><td><a href="https://github.com/microsoft/agent-framework/blob/11365dc3cb633256feae29772fd9a135b6211e9e/packages/core/agent_framework/_agents.py#L434">434</a>, <a href="https://github.com/microsoft/agent-framework/blob/11365dc3cb633256feae29772fd9a135b6211e9e/packages/core/agent_framework/_agents.py#L438">438</a>, <a href="https://github.com/microsoft/agent-framework/blob/11365dc3cb633256feae29772fd9a135b6211e9e/packages/core/agent_framework/_agents.py#L490">490</a>, <a href="https://github.com/microsoft/agent-framework/blob/11365dc3cb633256feae29772fd9a135b6211e9e/packages/core/agent_framework/_agents.py#L855">855</a>, <a href="https://github.com/microsoft/agent-framework/blob/11365dc3cb633256feae29772fd9a135b6211e9e/packages/core/agent_framework/_agents.py#L891">891</a>, <a href="https://github.com/microsoft/agent-framework/blob/11365dc3cb633256feae29772fd9a135b6211e9e/packages/core/agent_framework/_agents.py#L907">907</a>, <a href="https://github.com/microsoft/agent-framework/blob/11365dc3cb633256feae29772fd9a135b6211e9e/packages/core/agent_framework/_agents.py#L983-L986">983&ndash;986</a>, <a href="https://github.com/microsoft/agent-framework/blob/11365dc3cb633256feae29772fd9a135b6211e9e/packages/core/agent_framework/_agents.py#L1047-L1049">1047&ndash;1049</a>, <a href="https://github.com/microsoft/agent-framework/blob/11365dc3cb633256feae29772fd9a135b6211e9e/packages/core/agent_framework/_agents.py#L1170">1170</a>, <a href="https://github.com/microsoft/agent-framework/blob/11365dc3cb633256feae29772fd9a135b6211e9e/packages/core/agent_framework/_agents.py#L1186">1186</a>, <a href="https://github.com/microsoft/agent-framework/blob/11365dc3cb633256feae29772fd9a135b6211e9e/packages/core/agent_framework/_agents.py#L1188">1188</a>, <a href="https://github.com/microsoft/agent-framework/blob/11365dc3cb633256feae29772fd9a135b6211e9e/packages/core/agent_framework/_agents.py#L1201">1201</a>, <a href="https://github.com/microsoft/agent-framework/blob/11365dc3cb633256feae29772fd9a135b6211e9e/packages/core/agent_framework/_agents.py#L1207">1207</a>, <a href="https://github.com/microsoft/agent-framework/blob/11365dc3cb633256feae29772fd9a135b6211e9e/packages/core/agent_framework/_agents.py#L1243">1243</a>, <a href="https://github.com/microsoft/agent-framework/blob/11365dc3cb633256feae29772fd9a135b6211e9e/packages/core/agent_framework/_agents.py#L1245">1245</a>, <a href="https://github.com/microsoft/agent-framework/blob/11365dc3cb633256feae29772fd9a135b6211e9e/packages/core/agent_framework/_agents.py#L1254-L1259">1254&ndash;1259</a>, <a href="https://github.com/microsoft/agent-framework/blob/11365dc3cb633256feae29772fd9a135b6211e9e/packages/core/agent_framework/_agents.py#L1264">1264</a>, <a href="https://github.com/microsoft/agent-framework/blob/11365dc3cb633256feae29772fd9a135b6211e9e/packages/core/agent_framework/_agents.py#L1266">1266</a>, <a href="https://github.com/microsoft/agent-framework/blob/11365dc3cb633256feae29772fd9a135b6211e9e/packages/core/agent_framework/_agents.py#L1272-L1273">1272&ndash;1273</a>, <a href="https://github.com/microsoft/agent-framework/blob/11365dc3cb633256feae29772fd9a135b6211e9e/packages/core/agent_framework/_agents.py#L1280">1280</a>, <a href="https://github.com/microsoft/agent-framework/blob/11365dc3cb633256feae29772fd9a135b6211e9e/packages/core/agent_framework/_agents.py#L1282-L1283">1282&ndash;1283</a>, <a href="https://github.com/microsoft/agent-framework/blob/11365dc3cb633256feae29772fd9a135b6211e9e/packages/core/agent_framework/_agents.py#L1291-L1292">1291&ndash;1292</a>, <a href="https://github.com/microsoft/agent-framework/blob/11365dc3cb633256feae29772fd9a135b6211e9e/packages/core/agent_framework/_agents.py#L1295-L1297">1295&ndash;1297</a>, <a href="https://github.com/microsoft/agent-framework/blob/11365dc3cb633256feae29772fd9a135b6211e9e/packages/core/agent_framework/_agents.py#L1305-L1306">1305&ndash;1306</a>, <a href="https://github.com/microsoft/agent-framework/blob/11365dc3cb633256feae29772fd9a135b6211e9e/packages/core/agent_framework/_agents.py#L1308">1308</a>, <a href="https://github.com/microsoft/agent-framework/blob/11365dc3cb633256feae29772fd9a135b6211e9e/packages/core/agent_framework/_agents.py#L1313">1313</a>, <a href="https://github.com/microsoft/agent-framework/blob/11365dc3cb633256feae29772fd9a135b6211e9e/packages/core/agent_framework/_agents.py#L1315">1315</a></td></tr><tr><td><b>TOTAL</b></td><td><b>22180</b></td><td><b>2762</b></td><td><b>87%</b></td><td>&nbsp;</td></tr></tbody></table></details>

## Python Unit Test Overview
| Tests | Skipped | Failures | Errors | Time |
| ----- | ------- | -------- | -------- | ------------------ |
| 4677 | 247 :zzz: | 0 :x: | 0 :fire: | 1m 20s :stopwatch: |

**Comment by moonbox3** [comment]
Closing due to agent review errors.
